syntax = "proto2";

package proto;

service Predict {
  
  rpc PredictPerformance (PredictRequest) returns (PredictReply) {}
  rpc getModel (ModelRequest) returns (ModelReply) {}
  rpc trainModel (TrainRequest) returns (TrainReply) {}
}

message Any {
  oneof kind {
    int64 int_value = 1;
    float float_value = 2;
    string string_value = 3;
    ListOfStrings list_value = 4;
  }
}

message Pair {
    required string key = 1;
    required Any value = 2;
 }
 
 message Dictionary {
    //repeated Pair pairs = 1;
    map<string,Any> fields = 1;
 }

message PredictRequest {
  required string application = 1;
  required Dictionary features = 2;
  required string target = 3;
  required string variant = 4;
}


message PredictReply {
  required bool status = 1;
  required Dictionary results = 2;
  required Dictionary ml = 3;
  required string message = 4;
}

message ModelRequest {
  required string application = 1;
  required string target = 2;
}

message ModelReply {
  required Dictionary results = 1;
}

message ListValue {
  repeated ListOfStrings listOfStrings=1;
}

message ListOfStrings {
  repeated string strings=1;
}

message TrainRequest {
  required string application = 1;
  required string url_file = 2;
  required ListOfStrings features = 3;
  required string target = 4;
  required string variant = 5;
}

message TrainReply {
  required bool status = 1;
  required string message = 2;
  required string application = 3;
}